client/task_runner_test.go

```diff
@@ -195,3 +198,65 @@ func TestTaskRunner_SaveRestoreState(t *testing.T) {
 	}
 }
 */
+
+func TestTaskRunner_HandleDiscovery(t *testing.T) {
+	// Create the task runner
+	_, tr := testTaskRunner()
+
+	// Modify the job so we have predictable results
+	tr.allocID = "1"
+	tr.jobID = "job1"
+	tr.taskGroup = "group1"
+	tr.task.Name = "web"
+
+	// Assign dynamic ports. The ReservedPorts are overloaded for dynamic
+	// ports also, so we set those here as well.
+	tr.task.Resources.Networks[0].DynamicPorts = []string{"http", "https"}
+	tr.task.Resources.Networks[0].ReservedPorts = []int{80, 443}
+
+	// Create the mock discovery layer
+	providers := []discovery.Factory{discovery.NewMockDiscovery}
+	disc, err := discovery.NewDiscoveryLayer(providers, tr.config, tr.logger, tr.config.Node)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	tr.discovery = disc
+	mp := disc.Providers[0].(*discovery.MockDiscovery)
+
+	// Register the task with discovery
+	tr.handleDiscovery(false)
+
+	// Check the result
+	expect := map[string]int{
+		"job1.group1.web:1":       0,
+		"job1.group1.web.http:1":  80,
+		"job1.group1.web.https:1": 443,
```

Comment:
I like this approach, specially when using Docker host networking. (bind to host network namespace).  I would add ability to specify the ip address for registering the service at consul, and not only the port.

Imagine a situation where I have a server with multiple IPv4, and can spin up containers with --net=host and run my tasks by picking one of the ip addresses available at the HostOS network namespace.

This is possible already, since Nomad supports docker host network mode, so it would valuable to be able to customize address/port want to advertise to consul as part of our task.